addon/components/step-manager.js

```diff
@@ -189,25 +232,32 @@ export default Component.extend({
      */
     'transition-to-step'(to, value) {
       const from = get(this, 'transitions.currentStep');
+      const validator = this['will-transition'];
 
-      if (
-        this['will-transition'] &&
-        this['will-transition']({ value, from, to }) === false
-      ) {
+      // Prevent multiple transitions
+      if (get(this, 'loading')) {
         return;
       }
 
-      // Update the `currentStep` if it's mutable
-      if (this.attrs.currentStep && this.attrs.currentStep.update) {
-        this.attrs.currentStep.update(to);
-      }
+      if (validator && typeof validator === 'function') {
+        const result = validator({ value, from, to });
 
-      // Activate the next step
-      get(this, 'transitions').activate(to);
-
-      if (this['did-transition']) {
-        this['did-transition']({ value, from, to });
+        if (result && typeof result.then === 'function') {
+          set(this, 'loading', true);
+          result
+            .then(() => {
```

Comment:
I'm a bit concerned about the semantics around this, actually. I'm not sure that _any_ resolving promise should be the trigger for going to the next step, and _and_ failure rejecting it. Does it make more sense to instead resolve to `true` or `false`, and then handle a rejected promise the same as `false`? I think that might be better behavior...

Reply:
Well, I understand that while the promise resolve (let's say I got my info from a server), it might actually make me want to prevent the transition. According to you I should resolve returning `false` as value ? In my understanding, I will return `Promise.reject()`.

I think the right way to deal with such behavior is to return a `Promise.reject()` with an eventual `error` object as parameter (as advised [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject)).

Could you provide a use case where we can't use `Promise.reject()` to prevent the transition ?

N.B : we can also add a "logger" feature to handle the rejection for debugging purpose

Reply:
What I'm saying is that a resolving Promise doesn't always mean success. Think about the semantics of `fetch` -- a resolution does not mean the request was successful, and a rejection doesn't mean an error happen. A 400 or 500 HTTP code (which means an error happened) still resolve the `fetch` promise. Only an actual exception causes a `reject`. As such, I don't want to conflate a rejection (equivalent to an exception taking place) with the validation not passing an a resolution (no exception, but not necessarily a passing validation) with validation.

As for what an async validation action would actually resolve to, I think that maybe the best solution, rather than `true` or `false`, would be do expose a pair of `Symbol`s that we can match against.

Reply:
Okay I got your point. Regarding the `Symbol`, I think we should keep that simple. Also as a backward compatibility concern we should keep the `true` / `false`.